### PR TITLE
Fix race condition in concurrent schema deletion

### LINKED ISSUE #11605: Race condition in concurrent schema deletion
**Describe the bug**
In some scenarios, we deleted topics concurrently, which may trigger some race condition in deleting the same topic schema. For example, in `org.apache.pulsar.broker.service.AbstractTopic#deleteSchema` we will firstly perform `schemaRegistryService#getSchema`(read from bookkeeper) then `schemaRegistryService#deleteSchemaStorage`(delete its corresponding ledgers in bookkeeper). Therefore, when we delete a schema concurrently in two threads, it may happen that one thread has already deleted the ledger, and the other thread has just started executing and throws a "no such ledger" exception.

**To Reproduce**
1. start a cluster with at least two brokers
2. create a partitioned topic 'A' with multiple partitions
3. concurrently delete the non-partitioned topics within topic 'A' with the schema
4. then you will sporadically see 'No such ledger exists on Bookies' or 'NoNode for /schemas/XX/YY/ZZ' (Note that this phenomenon does not happen every time)

**Expected behavior**
Broker service can handle these scenarios correctly.

### PR DESCRIPTION
**Pull request by wuzhanpeng**
This PR may fixed #11605

### Motivation

Concurrently deleting topics with the same schema may cause race condition in broker side. If we do not handle these scenarios correctly we will get unexpected exceptions in broker logs.

### Modifications

1. Add existence checks before schema deletion in `AbstractTopic#deleteSchema`.
2. Add existence checks before actually performing schema storage deletion in `BookkeeperSchemaStorage#deleteSchema`.
3. Ignore `NoNodeException` in `BookkeeperSchemaStorage#deleteSchema`.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): no
  - The public API: no
  - The schema: no
  - The default values of configurations: no
  - The wire protocol: no
  - The rest endpoints: no
  - The admin cli options: no
  - Anything that affects deployment: no

### Documentation

This PR does not need to update the docs.

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java [hunk 1]
```diff
@@ -372,6 +374,21 @@ public String getReplicatorPrefix() {
         String id = TopicName.get(base).getSchemaName();
         SchemaRegistryService schemaRegistryService = brokerService.pulsar().getSchemaRegistryService();
         return schemaRegistryService.getSchema(id)
+                .exceptionally(t -> {
+                    if (t.getCause() != null
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:378]: This condition looks a little bit hacky.
1) we are only checking `t.getCause()`, IMHO we should have some utility method that traverses the full exception chain, looking for a SchemaException
2) a "non recoverable" SchemaException is not enough to say that this is a "Schema does not exist" exception, can we introduce some specific SchemaException subclass or method like `isRecoverable()` ?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/BookkeeperSchemaStorage.java [hunk 2]
```diff
@@ -370,7 +377,21 @@ public void close() throws Exception {
 
     @NotNull
     private CompletableFuture<Long> deleteSchema(String schemaId, boolean forceFully) {
-        return (forceFully ? CompletableFuture.completedFuture(null) : getSchema(schemaId))
+        return (forceFully ? CompletableFuture.completedFuture(null) : getSchema(schemaId).exceptionally(t -> {
+            if (t.getCause() != null
+                    && (t.getCause() instanceof SchemaException)
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/BookkeeperSchemaStorage.java:382]: same as above, at least we should have one utility method to detect this case

**Review by eolivelli — CHANGES_REQUESTED** [review]
is there already a test case in which we try to delete the Schema but the ledger on BookKeeper has already been deleted ? if not, can you please add it ?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java (hunk 1 above)
Comment (wuzhanpeng) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:378]: Thank you for your reply! 

For question 1, the condition is inspired by `SchemaRegistryServiceImpl#trimDeletedSchemaAndGetList(java.lang.String)`, that is, those schemas with unrecoverable exceptions would be regarded as deleted schemas. Therefore here I continue to use the design concept.

For question 2, I agree that unrecoverable SchemaException is not equivant to "Schema does not exist". However in schema storage deletion scenario, an unrecoverable exception, which includes `NoSuchLedgerException` and `NoSuchEntryException`, is enough to say that the schema has already been deleted by others and we can quit the process.

Comment (eolivelli) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java:378]: Thanks for your clarification 

What about creating an utility method and reduce code duplication?

**Review by eolivelli — APPROVED** [review]
LGTM

thank you

**Review by congbobo184 — CHANGES_REQUESTED** [review]
LGTM! could you please add some tests for it?

**Comment by wuzhanpeng** [comment]
> LGTM! could you please add some tests for it?

Sorry for the late. I supplemented a test case. PTAL~ @congbobo184

**Review by congbobo184 — APPROVED** [review]
LGTM!

**Review by sijie — APPROVED** [review]

**Pull request merged**